### PR TITLE
gh-105111: remove deprecated macros Py_TRASHCAN_SAFE_BEGIN and Py_TRASHCAN_SAFE_END

### LINKED ISSUE #105111: Remove deprecated Py_TRASHCAN_SAFE_BEGIN/ Py_TRASHCAN_SAFE_END
Py_TRASHCAN_SAFE_BEGIN/ Py_TRASHCAN_SAFE_END were [deprecated in 3.11](https://github.com/python/cpython/pull/27693/), to be replaced by Py_TRASHCAN_BEGIN/Py_TRASHCAN_END which were added in 3.8.   The deprecated macros are flawed (can lead to a crash) so we should not delay their removal. 

For more history, see https://mail.python.org/archives/list/python-dev@python.org/message/KWFX6XX3HMZBQ2BYBVL7G74AIOPWO66Y/.

### Linked PRs
* gh-105112

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #105111.

* Issue: gh-105111

----
:books: Documentation preview :books:: https://cpython-previews--105112.org.readthedocs.build/

### CODE REVIEW

File: Doc/whatsnew/3.13.rst [hunk 1]
```diff
@@ -281,6 +285,35 @@ Porting to Python 3.13
 This section lists previously described changes and other bugfixes
 that may require changes to your code.
 
+* The old trashcan macros (``Py_TRASHCAN_SAFE_BEGIN``/``Py_TRASHCAN_SAFE_END``)
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst]: I saw many times that removal and how to port removed code were documented at two different places. Why not move "how to port" in the Removed section?

File: Doc/whatsnew/3.13.rst [hunk 2]
```diff
@@ -273,6 +273,10 @@ Removed
   third-party Tix library which the module wrapped is unmaintained.
   (Contributed by Zachary Ware in :gh:`75552`.)
 
+* Remove the old trashcan macros (``Py_TRASHCAN_SAFE_BEGIN``/``Py_TRASHCAN_SAFE_END``).
+  are now deprecated. They should be replaced by the new macros ``Py_TRASHCAN_BEGIN``
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst]: I don't get the "are now deprecated" part. Mistake?

File: Doc/whatsnew/3.13.rst [hunk 3]
```diff
@@ -273,6 +273,10 @@ Removed
   third-party Tix library which the module wrapped is unmaintained.
   (Contributed by Zachary Ware in :gh:`75552`.)
 
+* Remove the old trashcan macros (``Py_TRASHCAN_SAFE_BEGIN``/``Py_TRASHCAN_SAFE_END``).
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst]: The macro names are important, i suggest to remove the parenthesis.

File: Doc/whatsnew/3.13.rst (hunk 1 above)
Comment (iritkatriel) [inline Doc/whatsnew/3.13.rst]:
```suggestion
* The old trashcan macros ``Py_TRASHCAN_SAFE_BEGIN`` and ``Py_TRASHCAN_SAFE_END``
```

File: Doc/whatsnew/3.13.rst (hunk 2 above)
Comment (iritkatriel) [inline Doc/whatsnew/3.13.rst]:
```suggestion
* Remove the old trashcan macros ``Py_TRASHCAN_SAFE_BEGIN`` and ``Py_TRASHCAN_SAFE_END``.
They should be replaced by the new macros ``Py_TRASHCAN_BEGIN``
```

File: Doc/whatsnew/3.13.rst (hunk 1 above)
Comment (iritkatriel) [reply Doc/whatsnew/3.13.rst]: I assume the idea is that the "porting" section tells you what might break when you try to run your code on the new version, so you can focus on that during the upgrade and ignore new features and bugifxes that aren't going to break your code.

File: Doc/whatsnew/3.13.rst [hunk 4]
```diff
@@ -273,6 +273,11 @@ Removed
   third-party Tix library which the module wrapped is unmaintained.
   (Contributed by Zachary Ware in :gh:`75552`.)
 
+* Remove the old trashcan macros ``Py_TRASHCAN_SAFE_BEGIN`` and
+  ``Py_TRASHCAN_SAFE_END``.  They should be replaced by the new macros
+  ``Py_TRASHCAN_BEGIN`` and ``Py_TRASHCAN_END``.  The new macros were
+  added in Python 3.8 and the old macros were deprecated in Python 3.11.i
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  added in Python 3.8 and the old macros were deprecated in Python 3.11.
```

File: Misc/NEWS.d/next/Core and Builtins/2023-05-30-20-30-57.gh-issue-105111.atn0_6.rst [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Remove the old trashcan macros
+(``Py_TRASHCAN_SAFE_BEGIN``/``Py_TRASHCAN_SAFE_END``). They should be
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2023-05-30-20-30-57.gh-issue-105111.atn0_6.rst]:
```suggestion
``Py_TRASHCAN_SAFE_BEGIN`` and ``Py_TRASHCAN_SAFE_END``. They should be
```

File: Misc/NEWS.d/next/Core and Builtins/2023-05-30-20-30-57.gh-issue-105111.atn0_6.rst [hunk 6]
```diff
@@ -0,0 +1,3 @@
...
+replaced by the new ≈macros ``Py_TRASHCAN_BEGIN`` and ``Py_TRASHCAN_END``.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2023-05-30-20-30-57.gh-issue-105111.atn0_6.rst]:
```suggestion
replaced by the new macros ``Py_TRASHCAN_BEGIN`` and ``Py_TRASHCAN_END``.
```

File: Misc/NEWS.d/next/Core and Builtins/2023-05-30-20-30-57.gh-issue-105111.atn0_6.rst [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Remove the old trashcan macros
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2023-05-30-20-30-57.gh-issue-105111.atn0_6.rst:1]: You may add "Patch by Irit Katriel." at the end ;-)

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for handling this one.

**Pull request merged**